lib/client/conntest/ssh.go

```diff
@@ -49,30 +86,231 @@ func NewSSHConnectionTester(cd services.ConnectionsDiagnostic) *SSHConnectionTes
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	key.TrustedCA = auth.AuthoritiesToTrustedCerts(certAuths)
+
+	keyAuthMethod, err := key.AsAuthMethod()
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	clusterName, err := s.cfg.UserClient.GetClusterName()
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	clientConfTLS, err := key.TeleportClientTLSConfig(nil, []string{clusterName.GetClusterName()})
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	key.KeyIndex = client.KeyIndex{
+		Username:    req.SSHPrincipal,
+		ProxyHost:   s.webProxyAddr,
+		ClusterName: clusterName.GetClusterName(),
+	}
+
+	host, err := s.getNodeHost(ctx, req.ResourceName)
+	if err != nil {
+		if trace.IsNotFound(err) {
+			connDiag, err := s.cfg.UserClient.AppendDiagnosticTrace(ctx, connectionDiagnosticID, types.NewTraceDiagnosticConnection(
+				types.ConnectionDiagnosticTrace_RBAC_NODE,
+				"Node not found. Ensure the Node exists and your role allows you to access it.",
+				err,
+			))
+			if err != nil {
+				return nil, trace.Wrap(err)
+			}
 
-	if err := s.cd.UpdateConnectionDiagnostic(ctx, connectionDiagnostic); err != nil {
+			return connDiag, nil
+		}
+
+		return nil, trace.Wrap(err)
+	}
+
+	processStdout := &bytes.Buffer{}
+
+	clientConf := client.MakeDefaultConfig()
+	clientConf.AddKeysToAgent = client.AddKeysToAgentNo
+	clientConf.AuthMethods = []ssh.AuthMethod{keyAuthMethod}
+	clientConf.Host = host
+	clientConf.HostKeyCallback = hostkeyCallback
+	clientConf.HostLogin = req.SSHPrincipal
+	clientConf.SkipLocalAuth = true
+	clientConf.SSHProxyAddr = s.sshProxyAddr
+	clientConf.Stderr = io.Discard
+	clientConf.Stdin = &bytes.Buffer{}
+	clientConf.Stdout = processStdout
+	clientConf.TLS = clientConfTLS
+	clientConf.TLSRoutingEnabled = s.cfg.TLSRoutingEnabled
+	clientConf.UseKeyPrincipals = true
+	clientConf.Username = currentUser.GetName()
+	clientConf.WebProxyAddr = s.webProxyAddr
+
+	tc, err := client.NewClient(clientConf)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	ctxWithTimeout, cancelFunc := context.WithTimeout(ctx, req.DialTimeout)
+	defer cancelFunc()
+
+	if err := tc.SSH(ctxWithTimeout, []string{"whoami"}, false); err != nil {
+		return s.handleErrFromSSH(ctx, connectionDiagnosticID, req.SSHPrincipal, err, processStdout)
+	}
+
+	connDiag, err := s.cfg.UserClient.AppendDiagnosticTrace(ctx, connectionDiagnosticID, types.NewTraceDiagnosticConnection(
+		types.ConnectionDiagnosticTrace_NODE_PRINCIPAL,
+		fmt.Sprintf("%q user exists in target node", req.SSHPrincipal),
+		nil,
+	))
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	connDiag.SetMessage(types.DiagnosticMessageSuccess)
+	connDiag.SetSuccess(true)
+
+	if err := s.cfg.UserClient.UpdateConnectionDiagnostic(ctx, connDiag); err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	return connDiag, nil
+}
+
+func (s SSHConnectionTester) getNodeHost(ctx context.Context, nodeName string) (host string, err error) {
+	if s.cfg.TLSRoutingEnabled {
```

Comment:
Same question as in https://github.com/gravitational/teleport/pull/15413#discussion_r959811730 

Why do we have a separate `getNodeHost` flow when TLS routing is enabled ?

Reply:
I removed the special flow and tested this manually
- cloud :heavy_check_mark: 
- local cluster with tls routing :heavy_check_mark: 
- local cluster without tls routing :heavy_check_mark: 

The issue was with the test.
I was not using the Node name after all.

I ended up removing the proxy client and the separate flow.

@smallinsky , can you please take another look?

Reply:
👍 Thanks for the explanation.

Reply:
Thank you for insisting on this :bow: